### PR TITLE
avoid cross shard receipts double spent 

### PR DESCRIPTION
**Pull request by chaosma**

### CODE REVIEW

File: core/rawdb/accessors_chain.go [hunk 1]
```diff
@@ -573,14 +573,12 @@ func ReadCXReceiptsProofUnspent(db DatabaseReader, shardID uint32, number uint64
...
 
-// TryWriteCXReceiptsProofUnspent check whether a CXReceiptsProof is unspent, write to database if unspent
-func TryWriteCXReceiptsProofUnspent(dbr DatabaseReader, dbw DatabaseWriter, shardID uint32, number uint64, blockHash common.Hash) error {
-	hash, _ := ReadCXReceiptsProofUnspent(dbr, shardID, number)
-	// only write to database for the first time given a specified block number
-	if hash == EmptyHash {
-		return dbw.Put(cxReceiptUnspentKey(shardID, number), blockHash.Bytes())
-	}
-	return ctxerror.New("[TryWriteCXReceiptsProofUnspent] blockHash already exist", "had", hash, "tryPut", blockHash)
+// WriteCXReceiptsProofUnspent check whether a CXReceiptsProof is unspent, write to database if unspent
```
Comment (rlan35) [inline core/rawdb/accessors_chain.go]: The code doesn’t check it’s unspent or not.

File: node/node_handler.go [hunk 2]
```diff
@@ -630,6 +631,10 @@ func (node *Node) PostConsensusProcessing(newBlock *types.Block) {
 			Msg("BINGO !!! Reached Consensus")
 	}
 
+	// mark incomingReceipts as spent and clean up
+	node.Blockchain().MarkCXReceiptsSpent(newBlock)
```
Comment (rlan35) [inline node/node_handler.go]: Should these be in the insert chain? So it’s more atomic operation?

File: core/blockchain.go [hunk 3]
```diff
@@ -2132,17 +2132,57 @@ func (bc *BlockChain) NextCXReceiptsProofUnpentCheckpoint(currentNum uint64, sha
...
 	for num := lastCheckpoint; num < newCheckpoint; num++ {
 		rawdb.DeleteCXReceiptsProofUnspent(bc.db, shardID, num)
 	}
 }
+
+// MarkCXReceiptsSpent mark the incomingReceipts in a block as spent
+// avoid double spent
+func (bc *BlockChain) MarkCXReceiptsSpent(block *types.Block) {
+	for _, cxp := range block.IncomingReceipts() {
+		rawdb.WriteCXReceiptsProofUnspent(bc.db, cxp)
```
Comment (rlan35) [inline core/blockchain.go]: This func should be named WriteCXReceiptsProofSpent, right?

Comment (chaosma) [reply core/blockchain.go]: My original thought is mark cxreceiptsproof as spent

File: node/node_handler.go (hunk 2 above)
Comment (chaosma) [reply node/node_handler.go]: en. good point, let me take a look

File: core/rawdb/accessors_chain.go (hunk 1 above)
Comment (chaosma) [reply core/rawdb/accessors_chain.go]: Yes, this is only write to database, the low level function. The checking is done in blockchain.go

Comment (chaosma) [reply core/rawdb/accessors_chain.go]: comments modified

File: node/node_handler.go (hunk 2 above)
Comment (chaosma) [reply node/node_handler.go]: move this into insertchain

**Pull request merged**